Phoenix/hello/lib/hello/catalog.ex

```diff
@@ -0,0 +1,219 @@
+defmodule Hello.Catalog do
+  @moduledoc """
+  The Catalog context.
+  """
+
+  import Ecto.Query, warn: false
+  alias Hello.Repo
+  alias Hello.Catalog.Category
+  alias Hello.Catalog.Product
+
+  @doc """
+  Returns the list of products.
+
+  ## Examples
+
+      iex> list_products()
+      [%Product{}, ...]
+
+  """
+  def list_products do
+    Repo.all(Product)
+  end
+
+  @doc """
+  Gets a single product.
+
+  Raises `Ecto.NoResultsError` if the Product does not exist.
+
+  ## Examples
+
+      iex> get_product!(123)
+      %Product{}
+
+      iex> get_product!(456)
+      ** (Ecto.NoResultsError)
+
+  """
+  def get_product!(id) do
+    Product |> Repo.get(id) |> Repo.preload(:categories)
+  end
+  @doc """
+  Creates a product.
+
+  ## Examples
+
+      iex> create_product(%{field: value})
+      {:ok, %Product{}}
+
+      iex> create_product(%{field: bad_value})
+      {:error, %Ecto.Changeset{}}
```

Comment:
@samu975 could you explain how a changeset works?

Reply:
changeset is used for type-casting, user input validation, and filtering parameters. Is a struct that stores a set of changes. We can use cast, validate_required, validate_format, validate_inclusion, etc.